### PR TITLE
Added a new "printConnectionTree" capability to AudioEngine and Node

### PR DESCRIPTION
**Pull request by btfranklin**
In AudioKit 4, there was a "printConnections" method on the main class. This was quite useful for debugging, though I found it a little bit difficult to mentally parse.

I found myself needing a somewhat similar ability in AudioKit 5, and I offer this approach for your consideration. A few things here:

1. I noticed that the concept of a "name" appears in several (though not all) Nodes. This is a useful commonality for debug output. To capture this, I added the NamedNode protocol.
2. I also noticed that there were a few different approaches to the initialization of name variables being used. I have made some updates to harmonize and align these, so the same approach is used consistently.
3. While the "printConnectionTree" method is useful from AudioEngine itself, there's no reason why it couldn't be applied to *any* node that can have inputs. For this reason, I added most of the functionality as an extension to Node itself, which relies on the aforementioned NamedNode protocol to provide richer output.

In practice, the use of this method produces quite nice output. Here is a real example from my project:
```
AudioKit |↳Mixer
AudioKit | ↳EqualizerFilter
AudioKit |  ↳EqualizerFilter
AudioKit |   ↳EqualizerFilter
AudioKit |    ↳EqualizerFilter
AudioKit |     ↳EqualizerFilter
AudioKit |      ↳EqualizerFilter
AudioKit |       ↳Compressor
AudioKit |        ↳Mixer
AudioKit |         ↳Mixer
AudioKit |          ↳Compressor
AudioKit |           ↳MIDISampler("Bass")
AudioKit |         ↳Mixer
AudioKit |          ↳Compressor
AudioKit |           ↳MIDISampler("Lead")
AudioKit |         ↳Mixer
AudioKit |          ↳Compressor
AudioKit |           ↳MIDISampler("Drone")
AudioKit |         ↳Mixer
AudioKit |          ↳DryWetMixer
AudioKit |           ↳MIDISampler("Drums")
AudioKit |           ↳ZitaReverb
AudioKit |            ↳MIDISampler("Drums")

```

I hope you agree that this is a useful addition to AudioKit 5.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -6,8 +6,8 @@ import PackageDescription
 let package = Package(
     name: "AudioKit",
     platforms: [
-        .macOS(.v10_14), .iOS(.v12), .tvOS(.v12)
-    ],
+        .macOS(.v10_14), .iOS(.v11), .tvOS(.v11)
+    ],
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Package.swift]: Fixed.

[the remark this answers is not in the document]
Comment (btfranklin) [reply Package.swift]: Fixed.

File: Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+extension AudioEngine {
+    
+    public func printConnectionTree() {
+        if let startNode = mainMixerNode {
+            startNode.printConnectionTree()
+        } else {
+            print("AudioKit | mainMixerNode is nil")
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift]: Fixed.

File: Sources/AudioKit/Internals/Utilities/AVAudioPCMBuffer+audition.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            Log(error, type: .error)
+            return
+        }
+        player.buffer = self
+        player.play()
+        sleep(self.frameCapacity / UInt32(self.format.sampleRate))
+        engine.stop()
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AVAudioPCMBuffer+audition.swift]: Fixed.

File: Sources/AudioKit/Internals/Engine/AudioEngine.swift [hunk 4]
```diff
@@ -50,9 +51,12 @@ public class AudioEngine {
     let _input = InputNode()
 
     /// Input for microphone or other device is created when this is accessed
-    public var input: InputNode {
-        guard let _ = Bundle.main.object(forInfoDictionaryKey: "NSMicrophoneUsageDescription") as? String else {
-            fatalError("To use the microphone, you must include the NSMicrophoneUsageDescription in your Info.plist")
+    public var input: InputNode? {
+        if #available(macOS 10.14, *) {
+            guard Bundle.main.object(forInfoDictionaryKey: "NSMicrophoneUsageDescription") != nil else {
+                Log("To use the microphone, you must include the NSMicrophoneUsageDescription in your Info.plist", type: .error)
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Engine/AudioEngine.swift]: Not my code.

File: Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift]: Fixed.

File: Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// Copyright AudioKit. All Rights Reserved. Revision History at http://github.com/AudioKit/AudioKit/
+
+import Foundation
+
+extension AudioEngine {
+
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift]: Fixed.

File: Sources/AudioKit/Internals/Utilities/AVAudioPCMBuffer+audition.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AVAudioPCMBuffer+audition.swift]: Fixed.

File: Sources/AudioKit/Internals/Engine/AudioEngine.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Engine/AudioEngine.swift]: Not my code.

File: Tests/AudioKitTests/Node Tests/Generator Tests/AudioPlayer2Tests.swift [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+        let pitchDuration = AUValue(duration) / AUValue(frequencies.count)
+
+        Log("duration", duration, "pitchDuration", pitchDuration)
+
+        let osc = Oscillator(waveform: Table(.square))
+        let engine = AudioEngine()
+        engine.output = osc
+
+        let url = FileManager.default.temporaryDirectory.appendingPathComponent("\(name)-\(AudioPlayer2Tests.tempFiles.count).aiff")
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/Generator Tests/AudioPlayer2Tests.swift]: Not my code.

File: Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (btfranklin) [reply Sources/AudioKit/Internals/Utilities/AudioEngine+printConnectionTree.swift]: Fixed.

File: Tests/AudioKitTests/Node Tests/Effects Tests/RhinoGuitarProcessorTests.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        let engine = AudioEngine()
+        let input = Oscillator()
+        engine.output = RhinoGuitarProcessor(input, preGain: 2.2)
+        input.start()
+        let audio = engine.startTest(totalDuration: 1.0)
+        audio.append(engine.render(duration: 1.0))
+        testMD5(audio)
+    }
+
+
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/Effects Tests/RhinoGuitarProcessorTests.swift]: Not my code.

File: Tests/AudioKitTests/Node Tests/EngineTests.swift [hunk 8]
```diff
@@ -0,0 +1,148 @@
+// Copyright AudioKit. All Rights Reserved. Revision History at http://github.com/AudioKit/AudioKit/
+@testable import AudioKit
+import AVFoundation
+import CAudioKit
+import XCTest
+
+// TODO: Needs engine inputNode tests
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/EngineTests.swift]: Not my code.

File: Tests/AudioKitTests/Node Tests/EngineTests.swift [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        let node1 = Mixer()
+        engine.output = node1
+
+        guard let mainMixerNode1 = engine.mainMixerNode else {
+            XCTFail("mainMixerNode1 wasn't created")
+            return
+        }
+        let mainMixerNodeSampleRate1 = mainMixerNode1.avAudioUnitOrNode.outputFormat(forBus: 0).sampleRate
+        XCTAssertTrue(mainMixerNodeSampleRate1 == audioFormat441k.sampleRate,
+                      "mainMixerNodeSampleRate is \(mainMixerNodeSampleRate1), requested rate was \(audioFormat441k.sampleRate)")
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/EngineTests.swift]: Not my code.

File: Tests/AudioKitTests/Node Tests/EngineTests.swift [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        let node2 = Mixer()
+        engine.output = node2
+
+        guard let mainMixerNode2 = engine.mainMixerNode else {
+            XCTFail("mainMixerNode2 wasn't created")
+            return
+        }
+        let mainMixerNodeSampleRate2 = mainMixerNode2.avAudioUnitOrNode.outputFormat(forBus: 0).sampleRate
+        XCTAssertTrue(mainMixerNodeSampleRate2 == audioFormat48k.sampleRate,
+                      "mainMixerNodeSampleRate2 is \(mainMixerNodeSampleRate2), requested rate was \(audioFormat48k.sampleRate)")
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/EngineTests.swift]: Not my code.

File: Tests/AudioKitTests/Node Tests/Generator Tests/AudioPlayer2Tests+Realtime.swift [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        player.isLooping = true
+        player.play()
+
+        wait(for: 5)
+        player.stop()
+    }
+
+    func realtimeInterrupts() {
```
[the remark this answers is not in the document]
Comment (btfranklin) [reply Tests/AudioKitTests/Node Tests/Generator Tests/AudioPlayer2Tests+Realtime.swift]: Not my code.

**Comment by aure** [comment]
Hi there. Thank you for being very diligent in your attention to detail.  There's one more modification I'd like you (or someone) to make on this PR.  Because this is adding a feature, I would like that feature to be accompanied with some tests so that if that feature is broken by future updates to AudioKit, we get notified. This seems like it should be very easy to do. Create a graph and then ensure that the output of printConnectionTree is an exact expected string.

**Comment by btfranklin** [comment]
@aure So, here's the thing: It's not really possible to directly test "print" outputs, so I had to really rethink how the code worked in order to make it testable...but this was a good thing! Thinking through how to develop testable code inspired me to reconsider some of the details of the approach, and I think the end result is substantially improved over the earlier iteration. Tests are now present from the Engine context as well as the Node context, and the tree is now a calculated variable. Please see my commit comments for deeper details.

**Comment by aure** [comment]
Enthusiastic thumbs up for the tests and new setup!

**Pull request merged**